## Import Necessary Libraries

In [1]:
import pandas as pd
import os

## Merge 12 months of sales data into a single CSV file

In [15]:
path = "./SalesAnalysis/Sales_Data/"
files = os.listdir(path)

get_csv = lambda file: pd.read_csv(f"{path}{file}")
get_csvs = lambda files: map(get_csv, files)
concat_csvs = lambda files: pd.concat(get_csvs(files))

concat_csvs(files).to_csv("all_data.csv", index = False)

## Understand the datastructure of the CSV

### Read in updated dataframe

In [16]:
all_data = pd.read_csv("all_data.csv")
all_data

Order ID                     Product Quantity Ordered Price Each  \
0        162009                      iPhone                1        700   
1        162009    Lightning Charging Cable                1      14.95   
2        162009            Wired Headphones                2      11.99   
3        162010  Bose SoundSport Headphones                1      99.99   
4        162011      34in Ultrawide Monitor                1     379.99   
...         ...                         ...              ...        ...   
186845   222905      AAA Batteries (4-pack)                1       2.99   
186846   222906            27in FHD Monitor                1     149.99   
186847   222907        USB-C Charging Cable                1      11.95   
186848   222908        USB-C Charging Cable                1      11.95   
186849   222909      AAA Batteries (4-pack)                1       2.99   

            Order Date                       Purchase Address  
0       03/28/19 20:59        942 Church St, Austin, TX 73301  
1       03/28/19 20:59        942 Church St, Austin, TX 73301  
2       03/28/19 20:59        942 Church St, Austin, TX 73301  
3       03/17/19 05:39   261 10th St, San Francisco, CA 94016  
4       03/10/19 00:01   764 13th St, San Francisco, CA 94016  
...                ...                                    ...  
186845  06/07/19 19:02          795 Pine St, Boston, MA 02215  
186846  06/01/19 19:29  495 North St, New York City, NY 10001  
186847  06/22/19 18:57  319 Ridge St, San Francisco, CA 94016  
186848  06/26/19 18:35   916 Main St, San Francisco, CA 94016  
186849  06/25/19 14:33         209 11th St, Atlanta, GA 30301  

[186850 rows x 6 columns]

### Print a concise summary of our dataframe

In [17]:
all_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186850 entries, 0 to 186849
Data columns (total 6 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Order ID          186305 non-null  object
 1   Product           186305 non-null  object
 2   Quantity Ordered  186305 non-null  object
 3   Price Each        186305 non-null  object
 4   Order Date        186305 non-null  object
 5   Purchase Address  186305 non-null  object
dtypes: object(6)
memory usage: 8.6+ MB


### Check the first few records

In [18]:
all_data.head()

Order ID                     Product Quantity Ordered Price Each  \
0   162009                      iPhone                1        700   
1   162009    Lightning Charging Cable                1      14.95   
2   162009            Wired Headphones                2      11.99   
3   162010  Bose SoundSport Headphones                1      99.99   
4   162011      34in Ultrawide Monitor                1     379.99   

       Order Date                      Purchase Address  
0  03/28/19 20:59       942 Church St, Austin, TX 73301  
1  03/28/19 20:59       942 Church St, Austin, TX 73301  
2  03/28/19 20:59       942 Church St, Austin, TX 73301  
3  03/17/19 05:39  261 10th St, San Francisco, CA 94016  
4  03/10/19 00:01  764 13th St, San Francisco, CA 94016

### Check the last few records

In [19]:
all_data.tail()

Order ID                 Product Quantity Ordered Price Each  \
186845   222905  AAA Batteries (4-pack)                1       2.99   
186846   222906        27in FHD Monitor                1     149.99   
186847   222907    USB-C Charging Cable                1      11.95   
186848   222908    USB-C Charging Cable                1      11.95   
186849   222909  AAA Batteries (4-pack)                1       2.99   

            Order Date                       Purchase Address  
186845  06/07/19 19:02          795 Pine St, Boston, MA 02215  
186846  06/01/19 19:29  495 North St, New York City, NY 10001  
186847  06/22/19 18:57  319 Ridge St, San Francisco, CA 94016  
186848  06/26/19 18:35   916 Main St, San Francisco, CA 94016  
186849  06/25/19 14:33         209 11th St, Atlanta, GA 30301

## Clean up the data

### Select all rows with NaN under an entire DataFrame
We just want to have an idea how many rows contain NaN values

In [7]:
all_data[all_data.isna().any(axis=1)]

Order ID Product Quantity Ordered Price Each Order Date  \
1047        NaN     NaN              NaN        NaN        NaN   
1542        NaN     NaN              NaN        NaN        NaN   
1653        NaN     NaN              NaN        NaN        NaN   
2014        NaN     NaN              NaN        NaN        NaN   
2043        NaN     NaN              NaN        NaN        NaN   
...         ...     ...              ...        ...        ...   
185795      NaN     NaN              NaN        NaN        NaN   
185868      NaN     NaN              NaN        NaN        NaN   
185887      NaN     NaN              NaN        NaN        NaN   
185960      NaN     NaN              NaN        NaN        NaN   
186580      NaN     NaN              NaN        NaN        NaN   

       Purchase Address  
1047                NaN  
1542                NaN  
1653                NaN  
2014                NaN  
2043                NaN  
...                 ...  
185795              NaN  
185868              NaN  
185887              NaN  
185960              NaN  
186580              NaN  

[545 rows x 6 columns]

### Create a function to drop rows with NaN values

In [8]:
drop_na = lambda df: df.dropna().reset_index(drop = True)
drop_na(all_data)

Order ID                     Product Quantity Ordered Price Each  \
0        162009                      iPhone                1        700   
1        162009    Lightning Charging Cable                1      14.95   
2        162009            Wired Headphones                2      11.99   
3        162010  Bose SoundSport Headphones                1      99.99   
4        162011      34in Ultrawide Monitor                1     379.99   
...         ...                         ...              ...        ...   
186300   222905      AAA Batteries (4-pack)                1       2.99   
186301   222906            27in FHD Monitor                1     149.99   
186302   222907        USB-C Charging Cable                1      11.95   
186303   222908        USB-C Charging Cable                1      11.95   
186304   222909      AAA Batteries (4-pack)                1       2.99   

            Order Date                       Purchase Address  
0       03/28/19 20:59        942 Church St, Austin, TX 73301  
1       03/28/19 20:59        942 Church St, Austin, TX 73301  
2       03/28/19 20:59        942 Church St, Austin, TX 73301  
3       03/17/19 05:39   261 10th St, San Francisco, CA 94016  
4       03/10/19 00:01   764 13th St, San Francisco, CA 94016  
...                ...                                    ...  
186300  06/07/19 19:02          795 Pine St, Boston, MA 02215  
186301  06/01/19 19:29  495 North St, New York City, NY 10001  
186302  06/22/19 18:57  319 Ridge St, San Francisco, CA 94016  
186303  06/26/19 18:35   916 Main St, San Francisco, CA 94016  
186304  06/25/19 14:33         209 11th St, Atlanta, GA 30301  

[186305 rows x 6 columns]

### Drop redundant header rows
For some reason, the raw data contains multiple header rows and this is causing problem when we try to convert our date column

Our data header is:
```
Order ID, Product, Quantity Ordered, Price Each, Order Date, Purchase Address
```

#### Create a function to get indexes of header rows
We need to know the indexes of the rows we want to drop from the dataframe

In [9]:
get_header_row_indexes = lambda df, col: df[df[col] == col].index
get_header_row_indexes(all_data, 'Order ID')

Int64Index([   433,    864,    930,   1979,   2032,   2107,   2485,   2728,
              3339,   4022,
            ...
            179246, 180955, 181479, 182134, 182146, 183671, 184012, 184041,
            184275, 186532],
           dtype='int64', length=355)

#### Create a function to drop rows with NaN values

In [10]:
drop_header_rows = lambda df, col='Order ID': df.drop(get_header_row_indexes(df, col))
drop_header_rows(all_data)

Order ID                     Product Quantity Ordered Price Each  \
0        162009                      iPhone                1        700   
1        162009    Lightning Charging Cable                1      14.95   
2        162009            Wired Headphones                2      11.99   
3        162010  Bose SoundSport Headphones                1      99.99   
4        162011      34in Ultrawide Monitor                1     379.99   
...         ...                         ...              ...        ...   
186845   222905      AAA Batteries (4-pack)                1       2.99   
186846   222906            27in FHD Monitor                1     149.99   
186847   222907        USB-C Charging Cable                1      11.95   
186848   222908        USB-C Charging Cable                1      11.95   
186849   222909      AAA Batteries (4-pack)                1       2.99   

            Order Date                       Purchase Address  
0       03/28/19 20:59        942 Church St, Austin, TX 73301  
1       03/28/19 20:59        942 Church St, Austin, TX 73301  
2       03/28/19 20:59        942 Church St, Austin, TX 73301  
3       03/17/19 05:39   261 10th St, San Francisco, CA 94016  
4       03/10/19 00:01   764 13th St, San Francisco, CA 94016  
...                ...                                    ...  
186845  06/07/19 19:02          795 Pine St, Boston, MA 02215  
186846  06/01/19 19:29  495 North St, New York City, NY 10001  
186847  06/22/19 18:57  319 Ridge St, San Francisco, CA 94016  
186848  06/26/19 18:35   916 Main St, San Francisco, CA 94016  
186849  06/25/19 14:33         209 11th St, Atlanta, GA 30301  

[186495 rows x 6 columns]

### Create the clean data variable and call all cleaning functions
The 2 functions we will be calling are `drop_na` and `drop_header_rows`.

We are expecting there will be `185950` rows left since there are `186850` original rows and `545` NaN values and `355` redundant header rows.

```
186850 - 545 - 355 = 185950 rows
```

In [11]:
def get_clean_data(df):
    return df.pipe(drop_na).pipe(drop_header_rows)
 
clean_data = get_clean_data(all_data)
clean_data

Order ID                     Product Quantity Ordered Price Each  \
0        162009                      iPhone                1        700   
1        162009    Lightning Charging Cable                1      14.95   
2        162009            Wired Headphones                2      11.99   
3        162010  Bose SoundSport Headphones                1      99.99   
4        162011      34in Ultrawide Monitor                1     379.99   
...         ...                         ...              ...        ...   
186300   222905      AAA Batteries (4-pack)                1       2.99   
186301   222906            27in FHD Monitor                1     149.99   
186302   222907        USB-C Charging Cable                1      11.95   
186303   222908        USB-C Charging Cable                1      11.95   
186304   222909      AAA Batteries (4-pack)                1       2.99   

            Order Date                       Purchase Address  
0       03/28/19 20:59        942 Church St, Austin, TX 73301  
1       03/28/19 20:59        942 Church St, Austin, TX 73301  
2       03/28/19 20:59        942 Church St, Austin, TX 73301  
3       03/17/19 05:39   261 10th St, San Francisco, CA 94016  
4       03/10/19 00:01   764 13th St, San Francisco, CA 94016  
...                ...                                    ...  
186300  06/07/19 19:02          795 Pine St, Boston, MA 02215  
186301  06/01/19 19:29  495 North St, New York City, NY 10001  
186302  06/22/19 18:57  319 Ridge St, San Francisco, CA 94016  
186303  06/26/19 18:35   916 Main St, San Francisco, CA 94016  
186304  06/25/19 14:33         209 11th St, Atlanta, GA 30301  

[185950 rows x 6 columns]

### Create a csv file for the clean data

In [12]:
clean_data.to_csv("clean_data.csv", index = False)

## Augment data with additional columns

### Add Month column

In [13]:
# Tutorial method
# all_data['Month'] = all_data['Order Date'].str[0:2]
# all_data['Month'] = all_data['Month'].astype('int32') # Need to use clean data since there are NaN values

In [14]:
get_order_month = lambda df: pd.to_datetime(df['Order Date'], format="%m/%d/%y %H:%M").dt.strftime('%m')

# Sidenote: Need to use the clean data since there are duplicate header rows which is causing error in date conversion 
clean_data.assign(
    Month = get_order_month(clean_data)) 

Order ID                     Product Quantity Ordered Price Each  \
0        162009                      iPhone                1        700   
1        162009    Lightning Charging Cable                1      14.95   
2        162009            Wired Headphones                2      11.99   
3        162010  Bose SoundSport Headphones                1      99.99   
4        162011      34in Ultrawide Monitor                1     379.99   
...         ...                         ...              ...        ...   
186300   222905      AAA Batteries (4-pack)                1       2.99   
186301   222906            27in FHD Monitor                1     149.99   
186302   222907        USB-C Charging Cable                1      11.95   
186303   222908        USB-C Charging Cable                1      11.95   
186304   222909      AAA Batteries (4-pack)                1       2.99   

            Order Date                       Purchase Address Month  
0       03/28/19 20:59        942 Church St, Austin, TX 73301    03  
1       03/28/19 20:59        942 Church St, Austin, TX 73301    03  
2       03/28/19 20:59        942 Church St, Austin, TX 73301    03  
3       03/17/19 05:39   261 10th St, San Francisco, CA 94016    03  
4       03/10/19 00:01   764 13th St, San Francisco, CA 94016    03  
...                ...                                    ...   ...  
186300  06/07/19 19:02          795 Pine St, Boston, MA 02215    06  
186301  06/01/19 19:29  495 North St, New York City, NY 10001    06  
186302  06/22/19 18:57  319 Ridge St, San Francisco, CA 94016    06  
186303  06/26/19 18:35   916 Main St, San Francisco, CA 94016    06  
186304  06/25/19 14:33         209 11th St, Atlanta, GA 30301    06  

[185950 rows x 7 columns]